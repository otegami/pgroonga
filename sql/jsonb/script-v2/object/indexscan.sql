CREATE TABLE logs (
  id int,
  record jsonb
);

INSERT INTO logs VALUES (1, '
{
  "tags": ["web"],
  "host": {"ipv4": "127.0.0.1"},
  "message": {"code": 100, "content": "hello"}
}
');
INSERT INTO logs VALUES (2, '
{
  "tags": ["mail"],
  "host": {"ipv4": "127.0.0.2"},
  "message": "hello"
}
');
INSERT INTO logs VALUES (3, '
{
  "tags": [],
  "host": {},
  "message": ["hello", "world"]
}
');

CREATE INDEX pgroonga_index ON logs
  USING pgroonga (record pgroonga_jsonb_ops_v2);

SET enable_seqscan = off;
SET enable_indexscan = on;
SET enable_bitmapscan = off;

EXPLAIN (COSTS OFF)
SELECT id, record
  FROM logs
 WHERE record &` 'paths @ ".message" && type == "object"'
 ORDER BY id;

SELECT id, record
  FROM logs
 WHERE record &` 'paths @ ".message" && type == "object"'
 ORDER BY id;

DROP TABLE logs;
