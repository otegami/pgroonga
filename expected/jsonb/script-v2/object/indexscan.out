CREATE TABLE logs (
  id int,
  record jsonb
);
INSERT INTO logs VALUES (1, '
{
  "tags": ["web"],
  "host": {"ipv4": "127.0.0.1"},
  "message": {"code": 100, "content": "hello"}
}
');
INSERT INTO logs VALUES (2, '
{
  "tags": ["mail"],
  "host": {"ipv4": "127.0.0.2"},
  "message": "hello"
}
');
INSERT INTO logs VALUES (3, '
{
  "tags": [],
  "host": {},
  "message": ["hello", "world"]
}
');
CREATE INDEX pgroonga_index ON logs
  USING pgroonga (record pgroonga_jsonb_ops_v2);
SET enable_seqscan = off;
SET enable_indexscan = on;
SET enable_bitmapscan = off;
EXPLAIN (COSTS OFF)
SELECT id, record
  FROM logs
 WHERE record &` 'paths @ ".message" && type == "object"'
 ORDER BY id;
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Sort
   Sort Key: id
   ->  Index Scan using pgroonga_index on logs
         Index Cond: (record &` 'paths @ ".message" && type == "object"'::text)
(4 rows)

SELECT id, record
  FROM logs
 WHERE record &` 'paths @ ".message" && type == "object"'
 ORDER BY id;
 id |                                             record                                             
----+------------------------------------------------------------------------------------------------
  1 | {"host": {"ipv4": "127.0.0.1"}, "tags": ["web"], "message": {"code": 100, "content": "hello"}}
(1 row)

DROP TABLE logs;
